### PR TITLE
Build Segments for User Schedule Segmentation

### PR DESCRIPTION
**Pull request by rdspring1**
## General Overview of Segmentation:

Segmentation decomposes a fusion into a directed acyclic graph (DAG) of sub-fusions. After applying the segmentation algorithm, we can translate the sub-fusions into their corresponding python definitions. Then, given the fusion's input arguments, the segments are run in the correct order to produce the output results.

The original FusionDefinition stores the sequence of sub-fusions and acts as an argument manager. It gathers the input arguments before running the sub-fusion and stores its results. To perform this function, it requires a map from the segment index space to the original index space. This mapping is generated while creating the python definition for each sub-fusion.

### CPP functions:
Step 1: `setupSegmentation` runs the segmentation algorithm on the CPP Fusion to create the `SegmentedFusion`. Then, sub-fusions are ordered according to their dependencies by the `prepareGroupOrder` function. It returns the number of segments in `SegmentedFusion`.

Step 2: `buildSegment` creates the CPP `Fusion` for a given segment id, translates it to a python `FusionDefinition`, then returns a mapping from the segment fusion state indices to the original fusion state indices.

Step 3: `finalizeSegmentation` destroys any state stored in `FusionDefinition`.

### Python functions:

1. `setupSegmentation`, `buildSegment`, and `finalizeSegmentation` are called together in `FusionDefinition.segment`.
6. If a python `FusionDefinition` has segments, call `_execute_segments` in the `FusionDefinition.execute`. The original `FusionDefinition` acts as argument manager, running the sub-fusions in topological order.

### Original Fusion: A reduction + broadcast + pointwise fusion.
```python
def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1],
                          contiguity=[True, True],
                          dtype=DataType.Float,
                          is_cpu=False)
    T1 = fd.define_tensor(shape=[-1, -1],
                          contiguity=[True, True],
                          dtype=DataType.Float,
                          is_cpu=False)
    T2 = fd.ops.sum(T0, dims=[1], keepdim=False, dtype=DataType.Float)
    T3 = fd.ops.broadcast(T2, is_broadcast_dim=[False, True])
    T4 = fd.ops.add(T1, T3)
    fd.add_output(T4)
```

### After Segmentation:
- The reduction scheduler does not support fusing any operations with an inner reduction, so the original fusion is divided into two segments. Segment 2 depends on Segment 1, so there is a strict ordering of the segments.
- The first segment contains the reduction and broadcast operations, which corresponds with [T0, T2, T3] in the original fusion. 
- The second segment is the pointwise addition with the broadcasted reduction. It corresponds with [T1, T3, T4] in the original fusion.

**First Segment:**
```python
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
   T0 = fd.define_tensor(shape=[-1, -1],
                         contiguity=[True, True],
                         dtype=DataType.Float,
                         is_cpu=False)
   T1 = fd.ops.sum(T0, dims=[1], keepdim=False, dtype=DataType.Float)
   T2 = fd.ops.broadcast(T1, is_broadcast_dim=[False, True])
   fd.add_output(T2)
```

**Second Segment:**
```python
def nvfuser_fusion_id3(fd : FusionDefinition) -> None :
   T0 = fd.define_tensor(shape=[-1, -1],
                         contiguity=[True, True],
                         dtype=DataType.Float,
                         is_cpu=False)
   T1 = fd.define_tensor(shape=[-1, 1],
                         contiguity=[True, None],
                         dtype=DataType.Float,
                         is_cpu=False)
   T2 = fd.ops.add(T0, T1)
   fd.add_output(T2)
```

## Changes in this PR

This PR implements `setupSegmentation` function for user-scheduler segmentation. It is the first PR in a stack, followed by https://github.com/NVIDIA/Fuser/pull/3335 and https://github.com/NVIDIA/Fuser/pull/3025.

1. Create `SegmentationState` class that contains all segmentation logic for python-frontend.
2. All segmentation logic is contained in a separate file - `csrc/python_frontend/segmentation.h`
3. `FusionDefinition` contains an instantiation of `SegmentationState` and exposes its logic in a public interface. This interface is added to the python bindings.
4. Created `test_segmentation_reduction_pointwise_epilogue` to test functionality.

### CODE REVIEW

**Comment by rdspring1** [comment]
!test

File: csrc/python_frontend/fusion_state.cpp [hunk 1]
```diff
@@ -68,6 +68,27 @@ std::ostream& operator<<(std::ostream& os, const State& state) {
   return os;
 }
 
+std::vector<Val*> getExtents(Fusion* fusion) {
```
Comment (rdspring1) [inline csrc/python_frontend/fusion_state.cpp:71]: Moved `getExtents` out of `FusionState`, so it is accessible in `SegmentationState`.

File: csrc/python_frontend/segmentation.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      std::move(segment_fusion_), &args, runtime_info);
+
+  // Get the order for fusion segments
+  prepareGroupOrder();
+
+  // Return the number of segments created by segmentation algorithm.
+  return (int64_t)segmented_fusion_->groups().size();
+}
+
+void SegmentationState::prepareGroupOrder() {
```
Comment (rdspring1) [inline csrc/python_frontend/segmentation.cpp:82]: `prepareGroupOrder` is like [prepareRuntimeOrder](https://github.com/NVIDIA/Fuser/blob/main/csrc/runtime/fusion_cache_utils.h#L42-L44).

Differences:
* It does not track `group_extent_binding_order`, which is the pe-determined order to bind tensor input metadata.

File: csrc/python_frontend/segmentation.h [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+#include <python_frontend/fusion_definition.h>
+#include <python_frontend/fusion_state.h>
+#include <runtime/fusion_executor_cache.h>
+
+namespace nvfuser::python_frontend {
+
+class FusionDefinition;
+
+//! Overview:
```
Comment (rdspring1) [inline csrc/python_frontend/segmentation.h:18]: I'd start reviewing from the description here.

File: csrc/python_frontend/segmentation.h [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+//! Segmentation decomposes a fusion into a directed acyclic graph (DAG) of
+//! sub-fusions. After applying the segmentation algorithm, we can translate
+//! the sub-fusions into their corresponding python definitions. Then, given the
+//! fusion's input arguments, the segments are run in the correct order to
+//! produce the output results.
+//!
+//! The original FusionDefinition stores the sequence of sub-fusions and acts as
+//! an argument manager. It gathers the input arguments before running the
+//! sub-fusion and stores its results. To perform this function, it requires a
+//! map from the segment index space to the original index space. This mapping
```
Comment (Priya2698) [inline csrc/python_frontend/segmentation.h:28]: IIUC, the segment index space and original index space refer to the input indices?
Can we specify that since we also have the indices for the segments/sub-fusions too?

File: csrc/python_frontend/segmentation.h [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+//!     S4 = fd.define_scalar(3, dtype=DataType.Int)
+//!     S5 = fd.define_scalar(100, dtype=DataType.Int32)
+//!     S6 = fd.define_scalar(1, dtype=DataType.Int32)
+//!     T7 = fd.ops.iota(S4, S5, S6, dtype=DataType.Int32)
+//!     fd.add_output(T3)
+//!     fd.add_output(T7)
+//!
+//! After Segmentation:
+//! The original fusion is divided into two segments. There is no dependencies
+//! between either segment so they can run in an order.
```
Comment (Priya2698) [inline csrc/python_frontend/segmentation.h:60]:
```Suggestion
//! between either segment so they can run in any order.
```

File: csrc/python_frontend/segmentation.cpp (hunk 2 above)
Comment (Priya2698) [reply csrc/python_frontend/segmentation.cpp:82]: naive question: Is it possible to just reuse `prepareRuntimeOrder` then?

File: csrc/python_frontend/segmentation.h [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  //  8)     Add SegmentedGroup to group_run_order_. Mark all outputs of
+  //         SegmentedGroup as ready.
+  //  9)   End For
+  //  10)  Fail if none of the SegmentedGroups are available to run.
+  //  11) End While
+  void prepareGroupOrder();
+
+ private:
+  // Clone of original fusion for segmentation
+  std::unique_ptr<Fusion> segment_fusion_ = nullptr;
```
Comment (Priya2698) [inline csrc/python_frontend/segmentation.h:203]: Would `cloned_fusion_` be a better name here, this is confusing since we also have `segmented_fusion_`?

File: csrc/python_frontend/segmentation.cpp (hunk 2 above)
Comment (Priya2698) [reply csrc/python_frontend/segmentation.cpp:82]: naive question: Is it possible then to reuse `prepareRuntimeOrder`?

File: csrc/python_frontend/segmentation.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      const std::vector<Val*>& group_outputs = group->outputs();
+      for (size_t group_out_i : c10::irange(group_outputs.size())) {
+        available_input.insert(group_outputs.at(group_out_i));
+      }
+      group_ran[group_i] = true;
+      ran_any_group = true;
+    }
+    NVF_ERROR(
+        ran_any_group,
+        "Failed to run all groups; An error must have occured in segmentation.");
```
Comment (Priya2698) [inline csrc/python_frontend/segmentation.cpp:142]:
```Suggestion
"Failed to run any groups; An error must have occured in segmentation.");
```

File: csrc/python_frontend/segmentation.h [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  //  1) Clone preschedFusion CPP Fusion.
+  //  2) Concretize fusion using input arguments.
+  //  3) Given the map_value_to_original_fid, the IRCloner returned by
+  //     Fusion::copy, AND symbolic_to_concrete map returned by concretization
+  //     pass, create a mapping from cloned Vals to original fusion state
+  //     indices
+  //  4) Get extents for cloned fusion
+  //  5) Create SchedulerRuntimeInfo
+  //  6) Run segmentation algorithm using cloned fusion, input arguments, and
+  //  scheduler runtime infomation.
```
Comment (jjsjann123) [inline csrc/python_frontend/segmentation.h:175]:
```suggestion
  //  scheduler runtime information.
```

File: csrc/python_frontend/segmentation.h [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  //  4) While not all the SegmentedGroups are run:
+  //  5)   For each SegmentedGroup:
+  //  6)     Skip SegmentedGroup if it is already run
+  //  7)     Skip SegmentedGroup if inputs are not ready
+  //  8)     Add SegmentedGroup to group_run_order_. Mark all outputs of
+  //         SegmentedGroup as ready.
+  //  9)   End For
+  //  10)  Fail if none of the SegmentedGroups are available to run.
+  //  11) End While
+  void prepareGroupOrder();
```
Comment (jjsjann123) [inline csrc/python_frontend/segmentation.h:199]: nitpick: does `prepareGroupOrder()` belong in the public? Looks like it's only used inside `setupSegmentation`

File: nvfuser/__init__.py [hunk 10]
```diff
@@ -54,6 +54,10 @@ def __init__(self, id=None, max_length=1024):
         super(FusionDefinition, self).__init__(id, max_length)
         self.profiled = False
 
+    def segment(self, inputs):
+        self.num_segments = self._setup_segmentation(inputs)
+        self._finalize_segmentation()
```
Comment (jjsjann123) [inline nvfuser/__init__.py:59]:
```suggestion
        num_segments = self._setup_segmentation(inputs)
        self._finalize_segmentation()
        return num_segments
```
nitpick: since segment isn't stateful. we probably don't want to keep `num_segments` inside self.

File: tests/python/test_python_frontend.py [hunk 11]
```diff
@@ -1244,6 +1244,28 @@ def fusion_func(fd: FusionDefinition):
...
+            fd.add_output(t3)
+
+        nvf_out1, _ = self.exec_nvfuser(fusion_func, inputs)
+        eager_out = torch.sum(inputs[0], dim=-1, keepdim=True) + inputs[1]
+        self.assertEqual(eager_out, nvf_out1[0])
+
+        with FusionDefinition() as fd:
+            fusion_func(fd)
+        fd.segment(inputs)
+        assert fd.num_segments == 2
```
Comment (jjsjann123) [inline tests/python/test_python_frontend.py:1267]:
```suggestion
        assert fd.segment(inputs) == 2
```

File: tests/python/test_python_frontend.py [hunk 12]
```diff
@@ -3881,7 +3903,7 @@ def fusion_func(fd: FusionDefinition) -> None:
 
             fd.add_output(T88)
 
-        nvf_out, _ = self.exec_nvfuser(fusion_func, inputs, is_clonable=True)
```
Comment (jjsjann123) [inline tests/python/test_python_frontend.py:3884]: IIUC, the default behavior is `is_clonable = True`, so this is just a clean up?!

File: csrc/python_frontend/fusion_definition.cpp [hunk 13]
```diff
@@ -673,4 +674,19 @@ std::vector<std::pair<double, double>> FusionDefinition::getValTolerances(
...
+      segmentation_state_ == nullptr, "SegmentationState already exists!");
+  segmentation_state_ = std::make_unique<SegmentationState>();
+  return segmentation_state_->setupSegmentation(
+      preschedFusion(), map_value_to_fid_, inputs);
+}
+
+void FusionDefinition::finalizeSegmentation() {
+  // Destroy SegmentedState
+  segmentation_state_.reset();
+}
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_definition.cpp:690]: a design question. What's the point of separating this into two steps?

Doesn't look like anything else is happening between `setupSegmentation()` to `finalizeSegmentation()`.

File: csrc/python_frontend/segmentation.cpp [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+void SegmentationState::prepareGroupOrder() {
+  NVF_ERROR(segmented_fusion_ != nullptr);
+
+  // Gather initial inputs for SegmentedFusion.
+  std::unordered_set<Val*> available_input;
+  std::copy(
+      segmented_fusion_->inputs().begin(),
+      segmented_fusion_->inputs().end(),
+      std::inserter(available_input, available_input.end()));
```
Comment (jjsjann123) [inline csrc/python_frontend/segmentation.cpp:90]: nitpick: is there any benefit from std::copy other than initializing directly with this?
```suggestion
  std::unordered_set<Val*> available_input(segmented_fusion_->inputs().begin(), segmented_fusion_->inputs().end());
```

File: nvfuser/__init__.py (hunk 10 above)
Comment (rdspring1) [reply nvfuser/__init__.py:59]: `num_segments` is used in the `test_segmentation_reduction_pointwise_epilogue`. We don't have anything else to check in this PR.

Comment (rdspring1) [reply nvfuser/__init__.py:59]: Ah, I can just check the return value.

File: csrc/python_frontend/fusion_definition.cpp (hunk 13 above)
Comment (rdspring1) [reply csrc/python_frontend/fusion_definition.cpp:690]: https://github.com/NVIDIA/Fuser/pull/3335 adds the `buildSegmentation()` to translate the CPP segments to their corresponding python definition.

We still need to destroy the `segmentation_state_` in this PR.

File: csrc/python_frontend/segmentation.h (hunk 4 above)
Comment (rdspring1) [reply csrc/python_frontend/segmentation.h:33]:
``` cpp
//! Each FusionDefinition contains a set of states representing tensors, vectors
//! and scalars. Every state has a unique index, which matches the insertion
//! order of the state in the FusionDefinition. These indices form a linear
//! index space for each FusionDefinition.
```

**Comment by rdspring1** [comment]
!test

File: csrc/python_frontend/segmentation.cpp (hunk 2 above)
Comment (rdspring1) [reply csrc/python_frontend/segmentation.cpp:82]: Maybe, but I prefer to keep them separate for now.

This is to be a jumping off point for improving the segmentation algorithm.

**Review by Priya2698 — APPROVED** [review]
LGTM.
